Add support for compensating temperature in Force/Torque sensors readings and for specify user offline offset

fjandrad opened the pull request:
Changes to be able to use the temperature as part of the correction performed by the secondary calibration matrix. It includes the rpc commands to use an offline estimated offset of the ft sensors.

fjandrad committed:
Adding functions to consider temperature coefficients when using the secondary calibration  matrix. Also added funciton to load temperature coefficients and the temperature offset using the tag FT_TEMPERATURE_COEFFICIENTS (it expects 7 values the offset is the last one)

fjandrad committed:
Merge branch 'master' of https://github.com/robotology/codyco-modules into origin/includeTemperatureInFT

fjandrad committed:
adding offset stuff

fjandrad committed:
adding vector for temperature measurments and mapping from ft position to temp array

fjandrad committed:
reading config and open functions ready, missing attaching function (if required ) and reading and using the temperature measurements. Still way to go for the constant offset

fjandrad committed:
using the temperature measurements ( still need to check if the attach is required) , also created usePreEstimatedOffset to use the offline estimated offset. Still need to check if the value is relevant to be used and avoid using garbage in case it was not set at configuration level

fjandrad committed:
changes to use MAS interfaces to get temperature. To open properly the MAS remapper

fjandrad committed:
changes to be able to launch wholebodydynamics externally for the iCubGenova04 that has 4 strain2 sensors and 2 strain1 sensors. Including a different prefix name, and the changes to use multipleanalogsensors

fjandrad committed:
changes to use the ids found in the actual strain of the sensor and to have both the configuration to launch with the robot and the ones to launch offline

fjandrad committed:
commit what is "working" for now. Only as external launch assuming the names of the sensors have and extra2 for the MAS sensor interfaces

fjandrad committed:
Solved the temperature and preEstimatedOffset funciton, might need some message cleaning

fjandrad committed:
Merge branch 'master' of https://github.com/robotology/codyco-modules into origin/includeTemperatureInFT

fjandrad committed:
changes to be compatible with yarp 3.2

fjandrad committed:
changes to add the use of permanent offset with rpc commands

fjandrad committed:
Changes to have usePreEstimated offset with rpc commands

fjandrad committed:
clean debugging messages

fjandrad committed:
Merge branch 'origin/includeTemperatureInFT' of https://github.com/robotology/codyco-modules into origin/includeTemperatureInFT

fjandrad commented:
It has been tested on the robot and it seems to be working. 

![axis1](https://user-images.githubusercontent.com/11043189/64547427-d3fb6300-d32c-11e9-9775-867b2c057079.png)
![axis2](https://user-images.githubusercontent.com/11043189/64547431-d3fb6300-d32c-11e9-8837-916a118e0316.png)
![axis3](https://user-images.githubusercontent.com/11043189/64547433-d3fb6300-d32c-11e9-9c4a-748802f63b5e.png)
![axis4](https://user-images.githubusercontent.com/11043189/64547434-d3fb6300-d32c-11e9-85bb-04a4ee7350d8.png)
![axis5](https://user-images.githubusercontent.com/11043189/64547435-d3fb6300-d32c-11e9-864b-88b2b9d30f47.png)
![axis6](https://user-images.githubusercontent.com/11043189/64547436-d493f980-d32c-11e9-8331-6c073d18d2a4.png)

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -188,6 +189,50 @@ class wholeBodyDynamicsDeviceFilters
+ *      </group>
+ * \endcode
+ *
+ * * * \subsection TemperatureCoefficients
```
This name perhaps need to be updated?

fjandrad replied:
Yes it should be offline estimated FT Offset

traversaro replied:
Friendly ping

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -347,6 +392,25 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
 
     /** F/T sensors interfaces */
     std::vector<yarp::dev::IAnalogSensor * > ftSensors;
+    //std::vector<yarp::dev::ISixAxisForceTorqueSensors * > ftSensors;
```
```suggestion
```

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -347,6 +392,25 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
+        //yarp::dev::IThreeAxisLinearAccelerometers * linearAccelerometers;
+//        yarp::dev::IOrientationSensors * orientationSensors;
+//        yarp::dev::IThreeAxisGyroscopes * gyroscopes;
+//        yarp::dev::IThreeAxisMagnetometers * magnetometers;
```
Can we remove this?

fjandrad replied:
I was thinking about leaving it in case someone wanted to extend it. Although from the current implementation of embObjFT and embObjfIMU they seem to be meant to be used always separated.  So we could.

fjandrad replied:
Removed

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -347,6 +392,25 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
+    double prevFTTempTimeStamp;
+
+//    /** Temperature sensors interfaces */
+//    std::vector<yarp::dev::ITemperatureSensors * > temperatureSensors;
```
```suggestion
```

traversaro commented on src/devices/wholeBodyDynamics/app/CMakeLists.txt:
```diff
@@ -82,6 +82,15 @@ install_wbd_robots_files(YARP_ROBOT_NAME iCubHeidelberg01
+#
+# iCubHeidelberg01 is one of a kind (a iCub with only legs and torso) and so it has specific configuration files
+#
+install_wbd_robots_files(YARP_ROBOT_NAME iCubGenova04
```
I am a bit confused by the comment connected with function, and in particula that this call will going to overwrite the files written in https://github.com/robotology/codyco-modules/pull/281/files#diff-4c6dd013f6e7d072ff01fd02e3ec8a42R60 .

fjandrad replied:
The comment is indeed wrong since it is talking about the previous code in line 80. This was created do to the fact that only iCubGenova04 and iCubGenova02 have strain2 sensors so they are the only ones that by default could use this configuration and we are not calibrating iCubGenova02 anytime soon to use temperature coefficients.  Another example of bad copy-paste-edit.

fjandrad replied:
Changed comment

traversaro commented on src/idl/wholeBodyDynamics_IDLServer/wholeBodyDynamics_IDLServer.thrift:
```diff
@@ -80,6 +80,13 @@ service wholeBodyDynamics_IDLServer
+   * @param calib_code argument to specify the sensors to reset (all,arms,legs,feet)
+   * @return true/false on success/failure
+   */
+  bool usePreEstimatedOffset(1:string calib_code)
```
The `calib_code` is a legacy parameter for the other methods. Given that `usePreEstimatedOffset` is a new method, perhaps we can just avoid introducing `calib_code` at all?

fjandrad replied:
I had a similar discussion with @prashanthr05, I was thinking of leaving it for the possibility of giving options later on such as selecting in which sensors to use it.  But I think that I will add it when I work for it and for now it might be better to not receive parameters.

fjandrad replied:
Done no parameters for now

traversaro commented:
The PR contains a lot of change, so I reviewed it but I  could have missed something. 
As this device is quite used, I think we need to make sure of the following: 
* Does the device work fine with an old configuration file,  such as the one currently present in robots-configuration? 
* Does the device still work fine if is connected to a FT sensor device for which temperature sensors are not supported (obviously disabling temperature compensation)? Unfortunately, we need to keep compatibility with iCub robots without temperature sensor that are the fast majority.

fjandrad committed:
Since we dont have log messages it is not required

It was meant to be temporary for the debug messages now that is gone it makes sense to take it off.

Co-Authored-By: Silvio Traversaro <silvio.traversaro@iit.it>

fjandrad commented on src/devices/wholeBodyDynamics/SixAxisForceTorqueMeasureHelpers.cpp:
```diff
@@ -9,19 +11,32 @@ SixAxisForceTorqueMeasureProcessor::SixAxisForceTorqueMeasureProcessor()
     toEigen(m_secondaryCalibrationMatrix).setIdentity();
     m_offset.zero();
+    toEigen(m_temperatureCoefficients).setZero();
+    m_tempOffset=0;
```
Sure I'll change it. I forget that for other languages ( not Matlab ) is better to put 0.0 for something that will be a float.

fjandrad replied:
Done

fjandrad committed:
Commented because it was not used

Remains from some first attempts should be deleted

Co-Authored-By: Silvio Traversaro <silvio.traversaro@iit.it>

fjandrad committed:
Another remain of previous tests

Not used so deleting.

Co-Authored-By: Silvio Traversaro <silvio.traversaro@iit.it>

fjandrad committed:
Update src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

Co-Authored-By: Silvio Traversaro <silvio.traversaro@iit.it>

fjandrad commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -509,7 +500,7 @@ bool WholeBodyDynamicsDevice::openExternalWrenchesPorts(os::Searchable& config)
         std::string port_name = outputWrenchPorts[i].port_name;
         outputWrenchPorts[i].output_port = new yarp::os::BufferedPort<yarp::sig::Vector>;
-        ok = ok && outputWrenchPorts[i].output_port->open(port_name);
+        ok = ok && outputWrenchPorts[i].output_port->open(portPrefix+port_name);
```
So a boolean to enable the use of portPrefix external wrench ports?
In theory portPrefix will only be used when more than one wholebodydynamics is running and in that case it was simpler to change just the port prefix. Otherwise for avoid port name collision you need to change not only the port prefix but also the individual names for all the external wrench ports. 
Having said that normally people do not launch more than one wholebodydynamics unless they are testing stuff like me.

fjandrad replied:
So given that you can fully change the name from the configuration file the boolean option might be worthless. So it would be just reverting to the previous right? That should keep it optional.

traversaro replied:
> 
> 
> So given that you can fully change the name from the configuration file the boolean option might be worthless. So it would be just reverting to the previous right? That should keep it optional.

The problem is that the `portPrefix` option is already used for other ports, so we can't  just change its default value to `""`, otherwise we will change the port name of a lot of other ports. The options are either: 
* have a boolean parameter with a name like "scopedExternalWrenchesPortNames` that enable or disable the use of `portPrefix` 
* have a different string parameter called `externalWrenchesPortPrefix` , that is just use for this port names
both are fine for me, the only thing that it would be not managable is to break compatibility with old port names, as those are already hardcoded in a lot of existing project/configuration files.

fjandrad replied:
> The problem is that the portPrefix option is already used for other ports, so we can't just change its default value to "", otherwise we will change the port name of a lot of other ports.

yeah I realized that when I looked in the code and started asking my self if I had introduced those changes as well haha. I think for now I will leave it as it was since is a change that is not related to the functionality of temperature compensation itself. We might open an issue and decide the best way to handle it and if its even necessary from the point of other users.

traversaro replied:
Ok for me.

fjandrad commented:
>     Does the device work fine with an old configuration file, such as the one currently present in robots-configuration?

yes, it does. For the comparison tests, I was actually running the currently present configuration also in my computer.

>   Does the device still work fine if it is connected to an FT sensor device for which temperature sensors are not supported (obviously disabling temperature compensation)? Unfortunately, we need to keep compatibility with iCub robots without temperature sensor that are the fast majority.

For using the features of this PR some actions are required:
1. If a sensor does not have temperature it should be specified in another file which is the configuration of the sensor.
1. For using the temperature compensation it has to be specifically named in the temperatureSensornames variable even if it is a multipleanalogsensor that has been named already in the SixAxisForceTorquSensorsNames variable.

So, in fact, to make it work with temperature you need extra steps, for sensors or robots that do not have this feature it should be transparent and no change required.

fjandrad committed:
Addressing reviewer suggested changes: 0 to 0.0 , erase commented code that is not useful, delete some debugging messages, change a comment that was wrongly copy-pasted, remove parameter dependency on the usePreEstimatedOffset rpc function

fjandrad committed:
Merge branch 'origin/includeTemperatureInFT' of https://github.com/robotology/codyco-modules into origin/includeTemperatureInFT

fjandrad commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -26,14 +26,14 @@ const size_t wholeBodyDynamics_nrOfChannelsOfAYARPIMUSensor = 12;
+    estimationWentWell(false),
+    validOffsetAvailable(false),
+    lastReadingSkinContactListStamp(0.0),
+    settingsEditor(settings)      
```
You are right I'll try to keep it clean next time.

traversaro commented:
@fjandrad I changed the title of the PR to be a bit more descriptive of the content of the PR

traversaro commented on src/devices/wholeBodyDynamics/SixAxisForceTorqueMeasureHelpers.cpp:
```diff
@@ -32,13 +46,41 @@ const iDynTree::Wrench& SixAxisForceTorqueMeasureProcessor::offset() const
+    iDynTree::Wrench ret;
+    fromEigen(ret,retEig);
+    return ret;
+}
```
As these is private code (their headers are not installed) if the variant of the methods **without temperature** is not used anymore, let's just remove them to simplify maintenance.

fjandrad replied:
Yes that seems to be the case, in the current code they are not used, should I directly eliminate them or do something like just send 0 in temperature so the behavior is the same while keeping the functionality even when calling the function without temperature?

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -1161,14 +1430,38 @@ void convertVectorFromDegreesToRadians(iDynTree::VectorDynSize & vector)
     {
         iDynTree::Wrench bufWrench;
         int ftRetVal = ftSensors[ft]->read(ftMeasurement);
+        if (timeFTStamp-prevFTTempTimeStamp>0.55){
```
Why 0.55 ? Can we make this a parameter loadable from configuration file with 0.55 as a default value?

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -2083,6 +2383,30 @@ bool WholeBodyDynamicsDevice::resetOffset(const std::string& calib_code)
+        double norm = iDynTree::toEigen(ftProcessors[ft].estimatedOffset()).norm();
+        if (norm!= 0.0){ // if norm is exactly 0.0 it means there was no offset  in the configuration file so it should be skipped
+
+            //The offset estimated offline is meant to be added, while it is substracted in the six axis force torque  measure helpers, so we changed the sign here. There might be a better solution though.
```
Can't we just change the convention of the FT offset parameter to be consistent with what is already there? These logic is actually quite old, and changing convention of FT offset just for some parameters is quite error while keeping the old one in other cases (for example when we print the offset) is extremely error prone.

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -2083,6 +2387,30 @@ bool WholeBodyDynamicsDevice::resetOffset(const std::string& calib_code)
+    for(size_t ft = 0; ft < this->getNrOfFTSensors(); ft++)
+    {
+        double norm = iDynTree::toEigen(ftProcessors[ft].estimatedOffset()).norm();
+        if (norm!= 0.0){ // if norm is exactly 0.0 it means there was no offset  in the configuration file so it should be skipped
```
We can just store a data structure that records if a sensor has a a-priori offset parameter if it is present in the `FT_OFFSET` configuration parameters, and just skip it if the a-priori offset is not there.

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp:
```diff
@@ -1161,14 +1430,38 @@ void convertVectorFromDegreesToRadians(iDynTree::VectorDynSize & vector)
+                remappedMASInterfaces.temperatureSensors->getTemperatureSensorName(ftTempMapping[ft],nameOfSensor);
+
 
+                if (sensorStatus==0 ){// this is yarp::dev::MAS_status MAS_OK                    
```
Instead of hardcoding 0, use directly `MAS_OK` so we are safe even if its value changes.

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -348,6 +393,19 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
+    yarp::sig::Vector ftTempMapping;
+    double prevFTTempTimeStamp;
+
+//    /** Temperature sensors interfaces */
```
Leftover?

fjandrad replied:
Yes sorry

traversaro commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -348,6 +393,19 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
+        yarp::dev::ISixAxisForceTorqueSensors * ftMultiSensors;
+        yarp::dev::IMultipleWrapper * multwrap;
+    } remappedMASInterfaces;
+    yarp::sig::Vector ftTempMapping;
```
If this is a int --> int mapping, can we just use a `std::vector<int>` or `std::vector<size_t>`?

fjandrad replied:
changed to std::vector<int>

traversaro commented on src/devices/wholeBodyDynamics/SixAxisForceTorqueMeasureHelpers.cpp:
```diff
@@ -32,13 +46,41 @@ const iDynTree::Wrench& SixAxisForceTorqueMeasureProcessor::offset() const
+iDynTree::Wrench& SixAxisForceTorqueMeasureProcessor::estimatedOffset()
+{
+    return m_estimated_offset;
+}
```
As this parameters is not used at all in this class, I suggest to move it outside of the other class, for example as a vector of iDynTree::Wrench in https://github.com/robotology/codyco-modules/pull/281/files#diff-a54e123540c7c6eb91d20ac92449e879R612 .

fjandrad replied:
Probably the wrong approach, but It made sense to me that each sensor processor keeps track of both the offset preEstimated or otherwise. 
Would it make more sense to do the offset replacement of https://github.com/robotology/codyco-modules/blob/0926d6f9ccb2a6d43a800a4daeb78a9e230062cb/src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp#L2399 inside the processor itself?

traversaro commented on src/idl/wholeBodyDynamics_IDLServer/wholeBodyDynamics_IDLServer.thrift:
```diff
@@ -80,6 +80,13 @@ service wholeBodyDynamics_IDLServer
+   * Use the offline estimated offset of the sensor.
+   * @param calib_code argument to specify the sensors to reset (all,arms,legs,feet)
+   * @return true/false on success/failure
+   */
```
Probably it could make sense to mention here that for sensor that do not have a pre-estimated offset, the offset is not change at all.

fjandrad replied:
Done

traversaro reviewed (changes requested):
A bit more comments. It would be great to test this version with the exiting conf  files in robots-configurations, so we are sure that we don't break anything.

traversaro commented:
Something that I think it would be really useful (but probably can go in a subsequent PR, as this one is already quite full of stuff) is to have an option to disable for some FT Sensors the initial offset calibration, and just use the pre-estimated offset without the need to call the RPC method.

fjandrad committed:
Remove portprefix from the external wrenches port

traversaro commented:
As there are a lot of conflicts in autogenerated files, can you rebase on top of the latest master? Probably you can just avoid manually fixing conflicts, and just regenerate the files after the rebase.

fjandrad commented on src/devices/wholeBodyDynamics/app/wholebodydynamics-icubgenova04-external.xml:
```diff
@@ -0,0 +1,114 @@
+        <action phase="shutdown" level="2" type="detach" />
+
+    </device>
+</devices>
```
Yeah. Any suggestions? 
Since is related to the temperature and the tests on iCubGenova04, I can think of:
- In the [issue ](https://github.com/dic-iit/element_ft-nonlinear-modelling/issues/10) in which I describe the datasets used for the paper and added as files needed to replicate the experiments. 
- In element [element_ft-nonlinear-modelling](https://github.com/dic-iit/element_ft-nonlinear-modelling).
- In the [insitu-ft-analysis repo](https://github.com/robotology-playground/insitu-ft-analysis/tree/master/docs/usedInRobot), which is public, and has the code to generate the temperature compensation. Although the code inside has dependency on iDyntree it does not neccessarily has anything to do with robot configuration files. I have also in the past put some there, but I don't think it fits.

fjandrad replied:
Or directly in robots-configuration?

fjandrad replied:
I put them in my local robots-configuration for now

fjandrad commented:
Apparently all the autogenerated files were deleted in master so I will rebase

traversaro commented:
> 
> 
> Apparently all the autogenerated files were deleted in master so I will rebase

I completely forgot about https://github.com/robotology/codyco-modules/pull/282 . I think you can just rebase and remove all the `autogenerated` directory.

fjandrad committed:
Changed description of function

fjandrad committed:
Merge branch 'master' of https://github.com/robotology/codyco-modules into origin/includeTemperatureInFT

fjandrad commented:
Actually I tried rebasing but started having conflicts with the commits between the filteredFT pull request and this pull request not sure why... So seeing it was more complicated, I pull from master and resolving those conflicts was easier for the merge since it was simply the removed files.

fjandrad commented on src/devices/wholeBodyDynamics/WholeBodyDynamicsDevice.h:
```diff
@@ -591,6 +662,13 @@ class WholeBodyDynamicsDevice :  public yarp::dev::DeviceDriver,
+     * @param calib_code argument to specify the sensors to reset (all,arms,legs,feet)
+     * @return true/false on success/failure
+     */
+    virtual bool usePreEstimatedOffset(const std::string& calib_code);
```
How does the override relates the thrift file and this file?

fjandrad replied:
I'll remove for now the input variables to the function in the device.

traversaro commented:
> Actually I tried rebasing but started having conflicts with the commits between the filteredFT pull request and this pull request not sure why...

Probably it would make sense to just squash all the commits in the PR to a one commit, and only after rebase on the top of master, otherwise you need to manually fix conflicts for each commit in this PR.

fjandrad commented on src/devices/wholeBodyDynamics/SixAxisForceTorqueMeasureHelpers.h:
```diff
@@ -57,17 +70,52 @@ class SixAxisForceTorqueMeasureProcessor
+    /**
+     * Process the input F/T.
+     *
+     * Returns secondaryCalibrationMatrix*input + offset .
```
True added the part related to temperature compensation

fjandrad committed:
Adressing requested changes: ftMapping to std::vector<int>, erasing the iCubGenova04 specific configuration files (probably moving to robots-config), correcting some documentation, remove the need for input variable in the preEstimatedOffset method in both .h and .cpp file

fjandrad commented on src/devices/wholeBodyDynamics/app/CMakeLists.txt:
```diff
@@ -82,6 +82,15 @@ install_wbd_robots_files(YARP_ROBOT_NAME iCubHeidelberg01
+install_wbd_robots_files(YARP_ROBOT_NAME iCubGenova04
+                         EXTERNAL_WBD_CONF wholebodydynamics-icubgenova04-external.xml
+                         EXTERNAL_YRI_CONF launch-wholebodydynamics-icubgenova04.xml
+                         ROBOT_WBD_CONF    wholebodydynamics-icub-eth-six-fts.xml)
```
Removed

fjandrad commented:
Given the changes I agree we need to verify on the robot again before merging.

traversaro commented:
This PR needs to be moved against whole-body-estimators due to https://github.com/robotology/whole-body-estimators/pull/16, however as there has been no code changes, it should be sufficient to just copy the update source code in the new repo.

fjandrad commented:
For the record the development of this feature was tracked in this feature: 
https://github.com/dic-iit/element_ft-nonlinear-modelling/issues/8

traversaro commented:
The updated version of this PR is provided in https://github.com/robotology/whole-body-estimators/pull/45, so we can close this PR.
